[ACRA 2.9] BasicImageFieldController onActivityResult Crash

I'll get on this tomorrow. Two issues: 

* Should be able to decode media
* An exception shouldn't crash the viewer.

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/879cca9f-143b-4db1-ad67-28831d6b223a

```
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=1, result=-1, data=Intent { dat=content://com.simplemobiletools.gallery.provider/external_files/data/data/com.simplemobiletools.gallery/files/storage/emulated/0/Pictures/Screenshots/[REDACTED] typ=image/png flg=0x1 }} to activity {com.ichi2.anki/com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity}: java.lang.IllegalStateException: Couldn't read row 0, col -1 from CursorWindow. Make sure the Cursor is initialized correctly before accessing data from it.
at android.app.ActivityThread.deliverResults(ActivityThread.java:4428)
at android.app.ActivityThread.handleSendResult(ActivityThread.java:4472)
at android.app.ActivityThread.-wrap19(Unknown Source:0)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1686)
at android.os.Handler.dispatchMessage(Handler.java:110)
at android.os.Looper.TinnoLoop(Looper.java:189)
at android.os.Looper.loop(Looper.java:270)
at android.app.ActivityThread.main(ActivityThread.java:6725)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
Caused by: java.lang.IllegalStateException: Couldn't read row 0, col -1 from CursorWindow. Make sure the Cursor is initialized correctly before accessing data from it.
at android.database.CursorWindow.nativeGetString(Native Method)
at android.database.CursorWindow.getString(CursorWindow.java:438)
at android.database.AbstractWindowedCursor.getString(AbstractWindowedCursor.java:51)
at android.database.CursorWrapper.getString(CursorWrapper.java:137)
at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.onActivityResult(BasicImageFieldController.java:6)
at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.onActivityResult(MultimediaEditFieldActivity.java:3)
at android.app.Activity.dispatchActivityResult(Activity.java:7403)
at android.app.ActivityThread.deliverResults(ActivityThread.java:4424)
... 10 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
@@ -68,7 +68,8 @@ import timber.log.Timber;
 
 public class BasicImageFieldController extends FieldControllerBase implements IFieldController {
 
-    private static final int ACTIVITY_SELECT_IMAGE = 1;
+    @VisibleForTesting
+    static final int ACTIVITY_SELECT_IMAGE = 1;
     private static final int ACTIVITY_TAKE_PICTURE = 2;
     private static final int IMAGE_SAVE_MAX_WIDTH = 1920;
 
@@ -225,8 +226,17 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
         }
         mImageFileSizeWarning.setVisibility(View.GONE);
         if (requestCode == ACTIVITY_SELECT_IMAGE) {
-            handleSelectImageIntent(data);
+            try {
+                handleSelectImageIntent(data);
+                mImageFileSizeWarning.setVisibility(View.GONE);
+            } catch (Exception e) {
+                AnkiDroidApp.sendExceptionReport(e, "BasicImageFieldController - handleSelectImageIntent");
+                Timber.e(e, "Failed to select image");
+                UIUtils.showThemedToast(mActivity, mActivity.getResources().getString(R.string.multimedia_editor_something_wrong), false);
+                return;
+            }
         } else if (requestCode == ACTIVITY_TAKE_PICTURE) {
+            mImageFileSizeWarning.setVisibility(View.GONE);
             String imagePath = rotateAndCompress(mTempCameraImagePath);
             mField.setImagePath(imagePath);
             mField.setHasTemporaryMedia(true);

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
@@ -36,6 +36,7 @@ import android.provider.MediaStore.MediaColumns;
 import androidx.annotation.VisibleForTesting;
 import androidx.core.content.FileProvider;
 
+import android.text.TextUtils;
 import android.text.format.Formatter;
 import android.util.DisplayMetrics;
 import android.view.Gravity;
@@ -232,7 +233,7 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
             } catch (Exception e) {
                 AnkiDroidApp.sendExceptionReport(e, "BasicImageFieldController - handleSelectImageIntent");
                 Timber.e(e, "Failed to select image");
-                UIUtils.showThemedToast(mActivity, mActivity.getResources().getString(R.string.multimedia_editor_something_wrong), false);
+                showSomethingWentWrong();
                 return;
             }
         } else if (requestCode == ACTIVITY_TAKE_PICTURE) {
@@ -248,18 +249,49 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
     }
 
 
+    private void showSomethingWentWrong() {
+        UIUtils.showThemedToast(mActivity, mActivity.getResources().getString(R.string.multimedia_editor_something_wrong), false);
+    }
+
+
     private void handleSelectImageIntent(Intent data) {
+        if (data == null) {
+            Timber.e("no intent provided");
+            showSomethingWentWrong();
+            return;
+        }
+
+        Timber.i("Handle Select Image. Intent: %s. extras: %s", data, data.getExtras() == null ? "null" : TextUtils.join(", ", data.getExtras().keySet()));
         Uri selectedImage = data.getData();
-        // Timber.d(selectedImage.toString());
+
+        if (selectedImage == null) {
+            Timber.w("selectedImage was null");
+            showSomethingWentWrong();
+            return;
+        }
+
         String[] filePathColumn = { MediaColumns.DATA };
 
         Cursor cursor = mActivity.getContentResolver().query(selectedImage, filePathColumn, null, null, null);
-        cursor.moveToFirst();
+
+        if (cursor == null) {
+            Timber.w("cursor was null");
+            showSomethingWentWrong();
+            return;
+        }
+
+        if (!cursor.moveToFirst()) {
+            //TODO: #5909, it would be best to instrument this to see if we can fix the failure
+            Timber.w("cursor had no data");
+            showSomethingWentWrong();
+            return;
+        }
 
         int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
         String filePath = cursor.getString(columnIndex);
         cursor.close();
 
+        Timber.i("Decoded image: '%s'", filePath);
         mField.setImagePath(filePath);
     }
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
@@ -223,18 +223,7 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
         }
         mImageFileSizeWarning.setVisibility(View.GONE);
         if (requestCode == ACTIVITY_SELECT_IMAGE) {
-            Uri selectedImage = data.getData();
-            // Timber.d(selectedImage.toString());
-            String[] filePathColumn = { MediaColumns.DATA };
-
-            Cursor cursor = mActivity.getContentResolver().query(selectedImage, filePathColumn, null, null, null);
-            cursor.moveToFirst();
-
-            int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
-            String filePath = cursor.getString(columnIndex);
-            cursor.close();
-
-            mField.setImagePath(filePath);
+            handleSelectImageIntent(data);
         } else if (requestCode == ACTIVITY_TAKE_PICTURE) {
             String imagePath = rotateAndCompress(mTempCameraImagePath);
             mField.setImagePath(imagePath);
@@ -242,6 +231,23 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
         }
         setPreviewImage(mField.getImagePath(), getMaxImageSize());
     }
+
+
+    private void handleSelectImageIntent(Intent data) {
+        Uri selectedImage = data.getData();
+        // Timber.d(selectedImage.toString());
+        String[] filePathColumn = { MediaColumns.DATA };
+
+        Cursor cursor = mActivity.getContentResolver().query(selectedImage, filePathColumn, null, null, null);
+        cursor.moveToFirst();
+
+        int columnIndex = cursor.getColumnIndex(filePathColumn[0]);
+        String filePath = cursor.getString(columnIndex);
+        cursor.close();
+
+        mField.setImagePath(filePath);
+    }
+
 
     @Override
     public void onFocusLost() {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
@@ -48,7 +48,9 @@ import android.widget.LinearLayout;
 import android.widget.LinearLayout.LayoutParams;
 import android.widget.TextView;
 
+import com.ichi2.anki.AnkiDroidApp;
 import com.ichi2.anki.R;
+import com.ichi2.anki.UIUtils;
 import com.ichi2.compat.CompatHelper;
 import com.ichi2.utils.BitmapUtil;
 import com.ichi2.utils.ExifUtil;
@@ -228,6 +230,9 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
             String imagePath = rotateAndCompress(mTempCameraImagePath);
             mField.setImagePath(imagePath);
             mField.setHasTemporaryMedia(true);
+        } else {
+            Timber.w("Unhandled request code: %d", requestCode);
+            return;
         }
         setPreviewImage(mField.getImagePath(), getMaxImageSize());
     }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivityTestBase.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivityTestBase.java
@@ -10,12 +10,15 @@ import com.ichi2.anki.multimediacard.fields.IField;
 import com.ichi2.anki.multimediacard.fields.IFieldController;
 import com.ichi2.anki.multimediacard.impl.MultimediaEditableNote;
 
+import org.mockito.Mockito;
 import org.robolectric.Robolectric;
 import org.robolectric.Shadows;
 import org.robolectric.android.controller.ActivityController;
 import org.robolectric.shadows.ShadowApplication;
 
 import androidx.test.core.app.ApplicationProvider;
+
+import static org.mockito.Mockito.when;
 
 public abstract class MultimediaEditFieldActivityTestBase extends RobolectricTest {
 
@@ -49,5 +52,12 @@ public abstract class MultimediaEditFieldActivityTestBase extends RobolectricTes
 
 
         return testCardTemplatePreviewer.getFieldController();
+    }
+
+    protected MultimediaEditFieldActivity setupActivityMock(IFieldController controller, MultimediaEditFieldActivity mActivity) {
+        MultimediaEditFieldActivity activity = Mockito.mock(MultimediaEditFieldActivity.class);
+        when(activity.getResources()).thenReturn(mActivity.getResources());
+        controller.setEditingActivity(activity);
+        return activity;
     }
 }

--- a/AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldControllerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldControllerTest.java
@@ -1,6 +1,13 @@
 package com.ichi2.anki.multimediacard.fields;
 
+import android.app.Activity;
+import android.content.ContentResolver;
+import android.content.Intent;
+
+import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity;
 import com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivityTestBase;
+import com.ichi2.testutils.AnkiAssert;
+import com.ichi2.testutils.MockContentResolver;
 
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -63,6 +70,24 @@ public class BasicImageFieldControllerTest extends MultimediaEditFieldActivityTe
         assertThat("A broken existing file should not display a preview",
                 controller.isShowingPreview(),
                 is(false));
+    }
+
+
+    @Test
+    public void invalidImageResultDoesNotCrashController() {
+        BasicImageFieldController controller = getValidControllerNoImage();
+        MultimediaEditFieldActivity activity = setupActivityMock(controller, controller.mActivity);
+
+        ContentResolver mock = MockContentResolver.returningEmptyCursor();
+        when(activity.getContentResolver()).thenReturn(mock);
+
+        //Act & Assert
+        AnkiAssert.assertDoesNotThrow(() -> performImageResult(controller, new Intent()));
+    }
+
+
+    private void performImageResult(BasicImageFieldController controller, Intent intent) {
+        controller.onActivityResult(BasicImageFieldController.ACTIVITY_SELECT_IMAGE, Activity.RESULT_OK, intent);
     }
 
     @CheckResult

--- a/AnkiDroid/src/test/java/com/ichi2/testutils/AnkiAssert.java
+++ b/AnkiDroid/src/test/java/com/ichi2/testutils/AnkiAssert.java
@@ -1,6 +1,7 @@
 package com.ichi2.testutils;
 
 import androidx.annotation.NonNull;
+import timber.log.Timber;
 
 import org.junit.Assert;
 
@@ -12,6 +13,7 @@ public class AnkiAssert {
         try {
             runnable.run();
         } catch (Exception e) {
+            Timber.e(e);
             Assert.fail();
         }
     }

--- a/AnkiDroid/src/test/java/com/ichi2/testutils/MockContentResolver.java
+++ b/AnkiDroid/src/test/java/com/ichi2/testutils/MockContentResolver.java
@@ -1,0 +1,24 @@
+package com.ichi2.testutils;
+
+import android.content.ContentResolver;
+import android.database.Cursor;
+
+import org.mockito.Mockito;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+public class MockContentResolver {
+
+    public static ContentResolver returningCursor(Cursor cursor) {
+        ContentResolver resolver = Mockito.mock(ContentResolver.class);
+        when(resolver.query(any(), any(), any(), any(), any())).thenReturn(cursor);
+        return resolver;
+    }
+
+    public static ContentResolver returningEmptyCursor() {
+        return returningCursor(MockCursor.getEmpty());
+    }
+
+
+}

--- a/AnkiDroid/src/test/java/com/ichi2/testutils/MockCursor.java
+++ b/AnkiDroid/src/test/java/com/ichi2/testutils/MockCursor.java
@@ -1,0 +1,20 @@
+package com.ichi2.testutils;
+
+import android.database.Cursor;
+import android.database.CursorWrapper;
+
+import org.mockito.Mockito;
+
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.Mockito.when;
+
+public class MockCursor {
+
+    public static Cursor getEmpty() {
+        //Couldn't just mock cursor
+        Cursor mockCursor = Mockito.mock(CursorWrapper.class);
+        when(mockCursor.moveToFirst()).thenReturn(false);
+        when(mockCursor.getString(anyInt())).thenThrow(new IllegalStateException());
+        return mockCursor;
+    }
+}

